### ISSUE TITLE
 Replication bootstrap stuck if interrupted

### ISSUE THREAD

**Issue by Mons** [tarantool/tarantool#5948]
If there was a mistake during replication bootstrap, bootstrap will never proceed after fixes.

**Steps to reproduce**

Run two instances in separace folders:
```sh
tarantool -e 'box.cfg{ listen=3301, replication={3301,3302} }'
```
```sh
tarantool -e 'box.cfg{ listen=3302, replication={3301,3302} }'
```

In this case `box.schema.user.grant` is forgotten, so both replicas are stuck in state (unimportant logs was cut):
```
main/103/interactive C> Tarantool 2.7.0-369-g97a927f51
main/103/interactive I> instance uuid 265a57b7-1620-440e-806b-c46b7ee52bbe
iproto/101/main I> binary: bound to [::]:3301
main/110/applier/ I> remote master 265a57b7-1620-440e-806b-c46b7ee52bbe at [::]:3301 running Tarantool 2.7.0
main/111/applier/ I> remote master 4596b545-d9ab-4c44-b7b7-94684af1e28a at [::]:3302 running Tarantool 2.7.0
main/103/interactive I> connected to 2 replicas
main/103/interactive I> initializing an empty data directory
main/103/interactive I> assigned id 1 to replica 265a57b7-1620-440e-806b-c46b7ee52bbe
main/103/interactive I> cluster uuid a1af0f3f-94ca-4455-88d1-04407e73200a
snapshot/101/main I> saving snapshot `./00000000000000000000.snap.inprogress'
snapshot/101/main I> done
main/111/applier/ I> can't join/subscribe
main/111/applier/ box.cc:2383 E> ER_LOADING: Instance bootstrap hasn't finished yet
main C> entering the event loop
main/111/applier/ I> can't connect to master
main/111/applier/ coio.cc:108 !> SystemError connect, called on fd 19, aka 127.0.0.1:60243: Connection refused
```
and
```
main/103/interactive C> Tarantool 2.7.0-369-g97a927f51
main/103/interactive I> instance uuid 4596b545-d9ab-4c44-b7b7-94684af1e28a
iproto/101/main I> binary: bound to [::]:3302
main/103/interactive I> connecting to 2 replicas
main/110/applier/ I> remote master 265a57b7-1620-440e-806b-c46b7ee52bbe at [::]:3301 running Tarantool 2.7.0
main/111/applier/ I> remote master 4596b545-d9ab-4c44-b7b7-94684af1e28a at [::]:3302 running Tarantool 2.7.0
main/103/interactive I> connected to 2 replicas
main/103/interactive I> bootstrapping replica from 265a57b7-1620-440e-806b-c46b7ee52bbe at [::]:3301
main/110/applier/ I> can't read row
main/110/applier/ box.cc:2269 E> ER_LOADING: Instance bootstrap hasn't finished yet
main/110/applier/ I> can't read row
main/110/applier/ session.cc:335 E> ER_ACCESS_DENIED: Read access to universe '' is denied for user 'guest'
```
If we inspect filesystem, we'll found, that there are snap and xlog under instance with lower uuid.

It's rather clear, what was the mistake, but we can't fix it, since after adding
```lua
box.schema.user.grant('guest','super',nil,nil, { if_not_exists = true  })
```
after `box.cfg` will not be executed anymore.

By hand solution is to remove those unbootstrapped  snap and xlog and re-run instances.
But it seems, that it could be done by tarantool itself.
Or at least tarantool may remove unfinished bootstrap files in case of termination.

**Comment by kokizzu** [comment]
Tried it with this:
```
box.cfg{
    listen = 13301,
      replication = {'replicator:password@127.0.0.1:13301',  -- master URI
                   'replicator:password@127.0.0.1:23301'}, -- replica URI
    read_only = false
  }
  box.once("schema", function()
     box.schema.user.create('replicator', {password = 'password'})
     box.schema.user.grant('replicator', 'replication') -- grant replication role
     box.schema.space.create("test")
     box.space.test:create_index("primary")
      
     print('box.once executed on master')
  end)
  box.schema.user.create('admin1', {password = 'admin1',if_not_exists=true}) -- added later
  box.schema.user.grant('admin1','read,write,execute,create,drop','universe')
```

and the replica:

```
box.cfg{
  listen = 23301,
  replication = {'replicator:password@127.0.0.1:13301',  -- master URI
                 'replicator:password@127.0.0.1:23301'}, -- replica URI
  read_only = true
}
box.once("schema", function()
   box.schema.user.create('replicator', {password = 'password'})
   box.schema.user.grant('replicator', 'replication') -- grant replication role
   box.schema.space.create("test")
   box.space.test:create_index("primary")
   print('box.once executed on replica 1')
end)
```

it could work

**Comment by sergepetrenko** [comment]
@Mons, thanks for the report!

Seems this is a misconfiguration, rather than a tarantool bug

Here's what's happening and how you can fix it:

1. On first cluster start you don't grant replication privileges, so the second node cannot bootstrap from the first one.
2. The first node creates a valid initial snapshot and is ready to serve requests (bootstrap doesn't require synchronizing with all nodes listed in `box.cfg.replication`, so the node becomes writeable right after snapshot creation)
3. You stop both nodes because you see the second node can't bootstrap from the first one
4. Now, on restart the first node is already "bootstrapped", so it enters "sync" state, where it tries to synchronize (subscribe and catch up) with everyone in `box.cfg.replication`. The booted node remains in that state for 5 minutes (`replication_sync_timeout` is 300 seconds by default), then becomes `orphan` and doesn't leave that state unless you reconfigure `box.cfg.replication`. This is why you can't issue a privilege grant on a second cluster start.

There are two ways out of that situation: first of all, the first node is already writeable right after step 2, so you may issue `box.schema.user.grant("guest", "replication")` in its console right after step 2. This way everything will work right away.

Now, if you prefer restarting both nodes, you have to add `replication_connect_quorum=1` or `replication_connect_quorum=0` to the first node's configuration. This way it will succeed in the 'sync' stage and become writeable. And your privilege grant will work just fine.

Note, you don't need to clear any bootstrap files, since the bootstrap isn't 'unfinished' after the first 2 steps. The bootstrap is done (in 1st node's point of view). The second node will catch up as soon as you grant necessary privileges to user 'guest'

**Long story short**, what you experienced was not a 'bootstrap hang', but an attempt to sync with a loading instance performed by the first node. In order to fix that, add `replication_connect_quorum=1` to the first node's configuration on restart. Or, alternatively, restart the first node without `replication` parameter, and add it after successful bootstrap of the second node.

**Comment by Mons** [comment]
Decided to left if off and to move to the complete removal of `replication_connect_quorum` and bundling it's behavior into the core
